File: 2023/ocaml/bin/day01.ml
open Core

let part_1_find_digit line =
  let rec aux = function
    | [] -> 0
    | x :: xs ->
      Char.is_digit x
      |> (function
       | true -> Char.to_int x - Char.to_int '0'
       | false -> aux xs)
  in
  aux line
;;

let part_1 lines =
  let rec aux = function
    | [] -> 0
    | x :: xs ->
      (part_1_find_digit (String.to_list x) * 10)
      + part_1_find_digit (String.to_list_rev x)
      + aux xs
  in
  aux lines
;;

let digits =
  [ "one", 1
  ; "two", 2
  ; "three", 3
  ; "four", 4
  ; "five", 5
  ; "six", 6
  ; "seven", 7
  ; "eight", 8
  ; "nine", 9
  ]
;;

let contains_digit str =
  List.find_map digits ~f:(fun (digit, value) ->
    match String.substr_index str ~pattern:digit with
    | Some _ -> Some value
    | None -> None)
;;

let part_2_find_digit line rev =
  let rec aux str = function
    | [] -> 0
    | x :: xs ->
      contains_digit str
      |> (function
       | Some value -> value
       | None ->
         Char.is_digit x
         |> (function
          | true -> Char.to_int x - Char.to_int '0'
          | false ->
            aux
              (match rev with
               | true -> String.append (Char.to_string x) str
               | false -> String.append str (Char.to_string x))
              xs))
  in
  aux "" line
;;

let part_2 lines =
  let rec aux = function
    | [] -> 0
    | x :: xs ->
      (part_2_find_digit (String.to_list x) false * 10)
      + part_2_find_digit (String.to_list_rev x) true
      + aux xs
  in
  aux lines
;;

let () =
  Aoc.command
    part_1
    part_2
    ~path:"../data/day1.txt"
    ~test_path:"../data/sample/day1part1.txt"
    ~test_path_2:"../data/sample/day1part2.txt"
    ~test_1_target:142
    ~test_2_target:281
    ()
  |> Command_unix.run
;;

File: 2025/python/day01.py
import unittest


def part1(path):
    val = 50
    res = 0
    with open(path, "r") as file:
        for line in file.readlines():
            i = int(line[1:])
            if line[0] == "L":
                i = -i
            val += i
            val = val % 100
            if val == 0:
                res += 1
    return res


def part2(path):
    val = 50
    res = 0
    with open(path, "r") as file:
        for line in file.readlines():
            i = int(line[1:])
            if line[0] == "L":
                i = -i
            val += i
            res += abs(val // 100)
            val = val % 100
    return res


class TestDay1(unittest.TestCase):
    def test_part1(self):
        self.assertEqual(part1("../data/sample/day01.txt"), 3)

    def test_part2(self):
        self.assertEqual(part2("../data/sample/day01.txt"), 6)


if __name__ == "__main__":
    print("part 1:", part1("../data/day01.txt"))
    print("part 2:", part2("../data/day01.txt"))

    unittest.main()
